### PR TITLE
`weakPointer` prototype implementation

### PR DESCRIPTION
**Pull request by jeremiah-corrado**
This is a prototype implementation for a `weakPointer` type. See https://github.com/chapel-lang/chapel/issues/20894 for some background.

So far, the interface is reasonably complete. I've implemented more methods than we'd probably want to keep in the final design. For example, there are currently two ways to construct a weakPointer from a shared class (`mySharedClass.downgrade()` and `new weakPointer(mySharedClass)`). We may not want both? We also might want to rename some of the methods/fields.

The implementation itself is mostly in place but there are some bugs yet to be worked out. (Neither of the tests are currently passing).

I'm creating a draft PR now to make it easier to collect feedback.

### CODE REVIEW

**Comment by vasslitvinov** [comment]
Do we want to provide borrowing from a weakPointer? Seems like we need it. Should borrowing upgrade the weakPointer while the borrowed pointer is alive, to ensure it remains valid?

Note to self: an optimization done in Boost https://stackoverflow.com/questions/5671241/how-does-weak-ptr-work

**Comment by jeremiah-corrado** [comment]
That's a good question. Do you have a scenario in mind where that would be useful?

In terms of implementation, I'm wondering if it would (a) create a temporary shared class instance and provide a borrow to it, or (b) provide the borrowed instance directly and increment the shared-reference count to keep it alive. Both options seem like they would be challenging for different reasons. With option a, wouldn't the shared class immediately go out of scope and get deinitialized? and with option (b), I think the borrowed class instance would have to know to decrement the shared reference count when it gets deinitialized (but only if it was initialized from a `weakPointer` to begin with).

**Comment by vasslitvinov** [comment]
> [Why do we need borrowing?]

The "normal" way to declare a method on a class results in the receiver having the borrowed type. [spec reference](https://chapel-lang.org/docs/main/language/spec/classes.html#class-methods)

While we could conceivably relax that for weak pointers, fundamentally we need borrowing to access fields. (@mppf how does this sound to you?)

> In terms of implementation, I'm wondering if it would (a) create a temporary shared class instance and provide a borrow to it, or (b) provide the borrowed instance directly and increment the shared-reference count to keep it alive. 

I would start with (a) because we can do that simply by inserting an implicit call to upgrade(). I am unsure about the lifetime of the resulting shared object. It is a situation akin to aliveness of the file object in `openFile().reader()`. So maybe it will "just work" ?

**Comment by vasslitvinov** [comment]
Another design question is whether we want ALL shared object to support weak pointers. I do not see how the other option -- have supportive and unsupportive flavors of shared -- could possibly work in practice. However, supporting weak pointers has cost in memory use and heap allocations even when WPs are not used. Are we willing to pay this cost?

Perhaps we could mitigate this by having the compiler check which classes use weak pointers and remove the overhead for all the other classes.

Meta-question: should we have these discussions on #20894?

**Comment by jeremiah-corrado** [comment]
> The "normal" way to declare a method on a class results in the receiver having the borrowed type.

The `weakPointer` type is implemented as a `record`. Is this still a concern for methods on records? (I don't think I've run into this issue in my implementation/testing so far)

> I am unsure about the lifetime of the resulting shared object. It is a situation akin to aliveness of the file object in openFile().reader(). So maybe it will "just work" ?

I'll have to test this to see if that would work, but I have a feeling the `shared` object would be dropped at the end of the `borrow()` method, leaving the borrowed object hanging.

> However, supporting weak pointers has cost in memory use and heap allocations even when WPs are not used.

I'm not too worried about additional overhead here because the changes to the `shared` implementation have been pretty minimal. The hidden `ReferenceCount` type is now carrying around one additional `atomic int`, and some of the operations on shared now require two atomic operations instead of one. This doesn't seem like it would degrade performance significantly but let me know if you feel differently.

> Meta-question: should we have these discussions on https://github.com/chapel-lang/chapel/issues/20894?

I think I'll open up a few issues next week for individual design questions after I get the prototype working. So this thread is probably ok for now.

**Comment by mppf** [comment]
I agree that the performance cost of having `shared` support weak pointers seems minimal. (But we should watch for a surprise there with benchmarks etc.)

In terms of "should one be able to borrow from a weak pointer", I think the answer is "no". The weak pointer might not be valid -- the object it refers to might no longer exist. What would the borrow do in that case? Throw? Return nil? I think it's more elegant to have whatever `upgrade` mechanism we have (which might be a cast or `myWeakPtr.upgrade()`) concern itself with that problem. And then if a borrow is needed, it can be safely done from the resulting `shared`.

File: modules/internal/SharedObject.chpl [hunk 1]
```diff
@@ -94,22 +94,42 @@ module SharedObject {
...
+      strongCount.add(1);
+      totalCount.add(1);
     }
     proc release() {
-      var oldValue = count.fetchSub(1);
+      var oldValue = strongCount.fetchSub(1);
+      return (oldValue - 1, totalCount.fetchSub(1) - 1);
+    }
+
+    proc tryRetainWeek(ref expected: int) {
```
Comment (mstrout) [inline modules/internal/SharedObject.chpl:115]: I think there is a misspelling here.

Comment (jeremiah-corrado) [reply modules/internal/SharedObject.chpl]: Thanks @mstrout! Should be fixed now.

**Pull request merged**